lib/backend/k8s/resources/node_conversion.go

```diff
@@ -15,43 +15,87 @@
 	kapiv1 "k8s.io/client-go/pkg/api/v1"
 )
 
+// Convert a Kubernetes format node, with Calico annotations, to a Calico Node
 func K8sNodeToCalico(node *kapiv1.Node) (*model.KVPair, error) {
-	// Get the internal IP of the node, should be 1-1 with calico node IP
-	nodeIP := ""
-	for _, address := range node.Status.Addresses {
-		if address.Type == kapiv1.NodeInternalIP {
-			nodeIP = address.Address
-			log.Debugf("Found NodeInternalIP %s", nodeIP)
-			break
-		}
-	}
-
-	ip := net.ParseIP(nodeIP)
-	log.Debugf("Node IP is %s", ip)
-	if ip == nil {
-		return nil, errors.New("Invalid IP received from k8s for Node")
-	}
-	asn := numorstring.ASNumber(64512)
-	return &model.KVPair{
+	kvp := model.KVPair{
 		Key: model.NodeKey{
 			Hostname: node.Name,
 		},
-		Value: &model.Node{
-			FelixIPv4:   ip,
-			Labels:      node.Labels,
-			BGPIPv4Addr: ip,
-			BGPIPv6Addr: &net.IP{},
-			BGPASNumber: &asn,
+		Value: model.Node{
+			Labels: node.Labels,
 		},
 		Revision: node.ObjectMeta.ResourceVersion,
-	}, nil
+	}
+
+	annotations := node.ObjectMeta.Annotations
+
+	cidrString, ok := annotations["projectcalico.org/IPv4Address"]; if ok {
+		ip, cidr, err := net.ParseCIDR(cidrString)
+		if err != nil {
+			log.Errorf("Could not parse BGPIPv4CIDR from k8s annotation: %s", cidrString)
+			return nil, err
+		}
+
+		kvp.Value.(model.Node).FelixIPv4   = ip
+		kvp.Value.(model.Node).BGPIPv4Addr = ip
+		kvp.Value.(model.Node).BGPIPv4Net  = cidr
+	}
+
+	asnString, ok := annotations["projectcalico.org/ASNumber"]; if ok {
+		asn, err := numorstring.ASNumberFromString(asnString)
+		if err != nil {
+			log.Errorf("Could not parse ASNumber from k8s annotation: %s", asnString)
+			return nil, err
+		}
+
+		kvp.Value.(model.Node).BGPASNumber = &asn
+	}
+
+	return &kvp, nil
+}
+
+// Convert a Calico Node to Kubernetes, place BGP configuration info in annotations
+func CalicoToK8sNode(kvp *model.KVPair) (*kapiv1.Node, error) {
+	calicoNode := kvp.Value.(model.Node)
+	annotations := map[string]string {
+		"projectcalico.org/IPv4Address": calicoNode.BGPIPv4Net.String(),
+		"projectcalico.org/ASNumber":    calicoNode.BGPASNumber.String(),
+	}
+
+	nodeMeta := kapiv1.ObjectMeta{
+		Name:        kvp.Key.(model.NodeKey).Hostname,
+		Annotations: annotations,
+		Labels:      calicoNode.Labels,
+	}
+
+	node := &kapiv1.Node{
+		ObjectMeta: nodeMeta,
+	}
+
+	return node, nil
+}
+
+// We take a k8s node and a Calico node and push the values from the Calico node into the k8s node
+func MakeK8sNode(kvp *model.KVPair, node *kapiv1.Node) (*kapiv1.Node, error) {
+	calicoNode := kvp.Value.(model.Node)
+	node.Annotations["projectcalico.org/IPv4Address"] = calicoNode.BGPIPv4Net.String()
+
+	// Don't set the ASNumber if it is nil
+	if calicoNode.BGPASNumber != nil {
+		node.Annotations["projectcalico.org/ASNumber"] = calicoNode.BGPASNumber.String()
+	}
+
+	// Add all of the calicoNode labels into the k8s node
+	for k, v := range calicoNode.Labels {
```

Comment:
This is a tiny bit scary.  I'm a little bit worried about us messing with labels on the underlying Node.

It's possible that it's the right thing to do, but also could lead to people unknowingly modifying their Nodes in ways they didn't expect.

On the other hand, users might expect to be able to modify the labels on a Node and have that change stick, so when it doesn't that could also be confusing.

WDYT?

CC @robbrockbank